lib/proxy/forward.c

```diff
@@ -484,6 +490,51 @@ static int forward_cmd_parse_dst(pool *p, const char *arg, char **name,
   return 0;
 }
 
+static int forward_cmd_parse_sni(pool *p, const struct proxy_conn **pconn) {
+  const char *sni = NULL;
+  char *hostport = NULL, *port_ptr = NULL, *uri = NULL;
+
+  sni = pr_table_get(session.notes, "mod_tls.sni", NULL);
+  if (sni == NULL) {
+    (void) pr_log_writefile(proxy_logfd, MOD_PROXY_VERSION,
+      "unable to use ProxyForwardMethod 'user@sni' due to missing TLS SNI");
+    errno = EPERM;
+    return -1;
+  }
+
+  port_ptr = strrchr(sni, ':');
```

Comment:
You definitely don't want that. SNI is used to match the certificate "CN". You can't add a port number there or the certificate validation will fail.

May I suggest either reusing the same port as the one the connection is made on and/or provide a configuration option for the port ?

Reply:
Technically, the SNI can be used by the TLS server to lookup the named resource however the server implementation sees fit.  The matching of a certificate Common Name field (as for a server certificate), or DNS SANs, occurs on the _client_ side of things.

In your particular setup, it may well be that the client will expect that the server certificate CN match the client-sent SNI -- but this is not necessarily true for _all_ cases.

In most servers, the hostname will not include a `:` followed by numbers, so I think the chances of a client using this `:<port-number>` syntax will be low.  If they do, then the behavior they see may not match, and it'll be addressed by having the client use the more expected hostname.

Thus I feel the inclusion of this optional bit to be useful for the regression suite, and low-enough impact/confusion for the real world as to not need a separate tunable setting for use.

Reply:
Ok fair enough. However I see it as a security risk since clients could connect on random ports and potentially do some probing around, It clearly needs to be documented that the user must restrict the port allowed in the config.

Reply:
The ability of clients to probe other ports unexpectedly, when forward proxying, is not specific to this SNI method; other `ProxyForwardMethod` mechanisms can be used for such probes as well.

That's why there the [`ProxyForwardTo`](https://htmlpreview.github.io/?https://github.com/Castaglia/proftpd-mod_proxy/blob/master/mod_proxy.html#ProxyForwardTo) regex restriction mechanism is provided.  It matches on destination host _and port_, and only allows forward connections to matching targets.  As such, it can be used to control/restrict those destination ports via SNI, just as it does for the other forward proxy methods.

I've updated this PR to show an example regex with port for the `ProxyForwardTo` docs, and adding a regression test for the SNI forward method showing that `ProxyForwardTo` blocks unexpected ports (as might be included in the SNI) as well.